File: minigit-3/py/diff_algo.py
"""Myers diff algorithm for minigit."""

from typing import List, Tuple, Optional
from dataclasses import dataclass


@dataclass
class DiffHunk:
    """A hunk in a diff."""
    old_start: int
    old_count: int
    new_start: int
    new_count: int
    lines: List[str]


def myers_diff(a: List[str], b: List[str]) -> List[Tuple[str, str]]:
    """
    Compute the diff between two lists of strings using Myers algorithm.
    Returns list of (op, line) where op is '+', '-', or ' '.
    """
    n = len(a)
    m = len(b)

    # Max edit distance
    max_d = n + m

    # V array for storing endpoints of furthest reaching paths
    v = {1: 0}

    # Trace for backtracking
    trace = []

    # Find the shortest edit script
    for d in range(max_d + 1):
        trace.append(dict(v))
        for k in range(-d, d + 1, 2):
            if k == -d or (k != d and v.get(k - 1, -1) < v.get(k + 1, -1)):
                x = v.get(k + 1, 0)
            else:
                x = v.get(k - 1, 0) + 1

            y = x - k

            # Extend diagonal
            while x < n and y < m and a[x] == b[y]:
                x += 1
                y += 1

            v[k] = x

            if x >= n and y >= m:
                # Found the end
                return backtrack(trace, a, b, d)

    return backtrack(trace, a, b, max_d)


def backtrack(trace: List[dict], a: List[str], b: List[str], d: int) -> List[Tuple[str, str]]:
    """Backtrack through trace to build the diff."""
    x = len(a)
    y = len(b)

    ops = []

    for d_idx in range(d, -1, -1):
        v = trace[d_idx]
        k = x - y

        if k == -d_idx or (k != d_idx and v.get(k - 1, -1) < v.get(k + 1, -1)):
            prev_k = k + 1
        else:
            prev_k = k - 1

        prev_x = v.get(prev_k, 0)
        prev_y = prev_x - prev_k

        # Add diagonal moves
        while x > prev_x and y > prev_y:
            x -= 1
            y -= 1
            ops.append((' ', a[x]))

        if d_idx > 0:
            if x == prev_x:
                # Insertion
                y -= 1
                ops.append(('+', b[y]))
            else:
                # Deletion
                x -= 1
                ops.append(('-', a[x]))

    ops.reverse()
    return ops


def compute_hunks(diff_ops: List[Tuple[str, str]], context: int = 3) -> List[DiffHunk]:
    """Group diff operations into hunks with context lines."""
    if not diff_ops:
        return []

    hunks = []
    current_hunk = None
    old_line = 0
    new_line = 0
    pending_context = []

    for op, line in diff_ops:
        if op == ' ':
            old_line += 1
            new_line += 1
            if current_hunk is not None:
                current_hunk.lines.append(f" {line}")
                current_hunk.old_count += 1
                current_hunk.new_count += 1
                # Check if we've had enough trailing context
                trailing_context = 0
                for l in reversed(current_hunk.lines):
                    if l.startswith(' '):
                        trailing_context += 1
                    else:
                        break
                if trailing_context >= context:
                    hunks.append(current_hunk)
                    current_hunk = None
                    pending_context = []
            else:
                pending_context.append((op, line, old_line, new_line))
                if len(pending_context) > context:
                    pending_context.pop(0)

        elif op == '-':
            if current_hunk is None:
                # Start new hunk with pending context
                old_start = old_line + 1 - len(pending_context)
                new_start = new_line + 1 - len(pending_context)
                current_hunk = DiffHunk(
                    old_start=old_start,
                    old_count=0,
                    new_start=new_start,
                    new_count=0,
                    lines=[]
                )
                for ctx_op, ctx_line, _, _ in pending_context:
                    current_hunk.lines.append(f" {ctx_line}")
                    current_hunk.old_count += 1
                    current_hunk.new_count += 1
                pending_context = []

            current_hunk.lines.append(f"-{line}")
            current_hunk.old_count += 1
            old_line += 1

        elif op == '+':
            if current_hunk is None:
                old_start = old_line + 1 - len(pending_context)
                new_start = new_line + 1 - len(pending_context)
                current_hunk = DiffHunk(
                    old_start=old_start,
                    old_count=0,
                    new_start=new_start,
                    new_count=0,
                    lines=[]
                )
                for ctx_op, ctx_line, _, _ in pending_context:
                    current_hunk.lines.append(f" {ctx_line}")
                    current_hunk.old_count += 1
                    current_hunk.new_count += 1
                pending_context = []

            current_hunk.lines.append(f"+{line}")
            current_hunk.new_count += 1
            new_line += 1

    if current_hunk is not None:
        # Trim trailing context to max context lines
        trailing = 0
        for i in range(len(current_hunk.lines) - 1, -1, -1):
            if current_hunk.lines[i].startswith(' '):
                trailing += 1
            else:
                break

        if trailing > context:
            to_remove = trailing - context
            current_hunk.lines = current_hunk.lines[:-to_remove]
            current_hunk.old_count -= to_remove
            current_hunk.new_count -= to_remove

        hunks.append(current_hunk)

    return hunks


def format_diff(old_path: str, new_path: str, hunks: List[DiffHunk],
                old_label: str = 'a', new_label: str = 'b') -> str:
    """Format diff hunks as unified diff output."""
    if not hunks:
        return ""

    lines = []
    lines.append(f"diff --git {old_label}/{old_path} {new_label}/{new_path}")
    lines.append(f"--- {old_label}/{old_path}")
    lines.append(f"+++ {new_label}/{new_path}")

    for hunk in hunks:
        old_range = f"{hunk.old_start},{hunk.old_count}" if hunk.old_count != 1 else str(hunk.old_start)
        new_range = f"{hunk.new_start},{hunk.new_count}" if hunk.new_count != 1 else str(hunk.new_start)
        lines.append(f"@@ -{old_range} +{new_range} @@")
        lines.extend(hunk.lines)

    return '\n'.join(lines)


def diff_files(old_content: Optional[str], new_content: Optional[str],
               old_path: str, new_path: str = None,
               context: int = 3) -> str:
    """
    Compute unified diff between two file contents.
    Returns empty string if no changes.
    """
    if new_path is None:
        new_path = old_path

    if old_content is None:
        old_lines = []
    else:
        old_lines = old_content.splitlines()

    if new_content is None:
        new_lines = []
    else:
        new_lines = new_content.splitlines()

    if old_lines == new_lines:
        return ""

    diff_ops = myers_diff(old_lines, new_lines)
    hunks = compute_hunks(diff_ops, context)

    return format_diff(old_path, new_path, hunks)


def diff_binary() -> str:
    """Return message for binary file diff."""
    return "Binary files differ"

File: minigit-3/ts/src/diff-algo.ts
// Myers diff algorithm implementation

import { DiffHunk, FileDiff } from './types';

interface Edit {
  type: 'insert' | 'delete' | 'equal';
  oldLine?: string;
  newLine?: string;
  oldIndex: number;
  newIndex: number;
}

// Myers diff algorithm
export function myersDiff(oldLines: string[], newLines: string[]): Edit[] {
  const n = oldLines.length;
  const m = newLines.length;
  const max = n + m;
  const v: { [k: number]: number } = { 1: 0 };
  const trace: { [k: number]: number }[] = [];

  for (let d = 0; d <= max; d++) {
    trace.push({ ...v });

    for (let k = -d; k <= d; k += 2) {
      let x: number;
      if (k === -d || (k !== d && v[k - 1] < v[k + 1])) {
        x = v[k + 1];
      } else {
        x = v[k - 1] + 1;
      }

      let y = x - k;

      while (x < n && y < m && oldLines[x] === newLines[y]) {
        x++;
        y++;
      }

      v[k] = x;

      if (x >= n && y >= m) {
        return backtrack(trace, oldLines, newLines);
      }
    }
  }

  return [];
}

function backtrack(trace: { [k: number]: number }[], oldLines: string[], newLines: string[]): Edit[] {
  const edits: Edit[] = [];
  let x = oldLines.length;
  let y = newLines.length;

  for (let d = trace.length - 1; d >= 0; d--) {
    const v = trace[d];
    const k = x - y;

    let prevK: number;
    if (k === -d || (k !== d && (v[k - 1] ?? -1) < (v[k + 1] ?? -1))) {
      prevK = k + 1;
    } else {
      prevK = k - 1;
    }

    const prevX = v[prevK] ?? 0;
    const prevY = prevX - prevK;

    while (x > prevX && y > prevY) {
      x--;
      y--;
      edits.unshift({
        type: 'equal',
        oldLine: oldLines[x],
        newLine: newLines[y],
        oldIndex: x,
        newIndex: y
      });
    }

    if (d > 0) {
      if (x === prevX) {
        y--;
        edits.unshift({
          type: 'insert',
          newLine: newLines[y],
          oldIndex: x,
          newIndex: y
        });
      } else {
        x--;
        edits.unshift({
          type: 'delete',
          oldLine: oldLines[x],
          oldIndex: x,
          newIndex: y
        });
      }
    }
  }

  return edits;
}

export function createHunks(edits: Edit[], contextLines: number = 3): DiffHunk[] {
  const hunks: DiffHunk[] = [];
  let currentHunk: DiffHunk | null = null;
  let lastChangeIndex = -1;

  for (let i = 0; i < edits.length; i++) {
    const edit = edits[i];

    if (edit.type !== 'equal') {
      // Start a new hunk or extend current one
      if (currentHunk === null) {
        // Start new hunk with context before
        const contextStart = Math.max(0, i - contextLines);
        currentHunk = {
          oldStart: edits[contextStart].oldIndex + 1, // 1-indexed
          oldCount: 0,
          newStart: edits[contextStart].newIndex + 1, // 1-indexed
          newCount: 0,
          lines: []
        };

        // Add context before
        for (let j = contextStart; j < i; j++) {
          currentHunk.lines.push(' ' + (edits[j].oldLine || edits[j].newLine || ''));
          currentHunk.oldCount++;
          currentHunk.newCount++;
        }
      }

      // Add the change
      if (edit.type === 'delete') {
        currentHunk.lines.push('-' + (edit.oldLine || ''));
        currentHunk.oldCount++;
      } else if (edit.type === 'insert') {
        currentHunk.lines.push('+' + (edit.newLine || ''));
        currentHunk.newCount++;
      }

      lastChangeIndex = i;
    } else if (currentHunk !== null) {
      // Equal line after a change
      const distanceToLastChange = i - lastChangeIndex;

      if (distanceToLastChange <= contextLines * 2) {
        // Still within context, add as context line
        currentHunk.lines.push(' ' + (edit.oldLine || ''));
        currentHunk.oldCount++;
        currentHunk.newCount++;
      } else {
        // Context exceeded, check if we should start a new hunk
        // First, close current hunk
        hunks.push(currentHunk);
        currentHunk = null;
      }
    }
  }

  // Close any remaining hunk
  if (currentHunk !== null) {
    // Trim trailing context to contextLines
    const lastChange = currentHunk.lines.length - 1;
    let trailingContext = 0;
    for (let i = lastChange; i >= 0; i--) {
      if (currentHunk.lines[i].startsWith(' ')) {
        trailingContext++;
      } else {
        break;
      }
    }

    if (trailingContext > contextLines) {
      const excess = trailingContext - contextLines;
      currentHunk.lines = currentHunk.lines.slice(0, -excess);
      currentHunk.oldCount -= excess;
      currentHunk.newCount -= excess;
    }

    hunks.push(currentHunk);
  }

  return hunks;
}

export function formatDiff(fileDiff: FileDiff): string {
  const lines: string[] = [];

  // Header
  lines.push(`diff --git a/${fileDiff.oldPath} b/${fileDiff.newPath}`);

  if (fileDiff.isNew) {
    lines.push('new file mode ' + (fileDiff.newMode || '100644'));
  } else if (fileDiff.isDeleted) {
    lines.push('deleted file mode ' + (fileDiff.oldMode || '100644'));
  } else if (fileDiff.oldMode !== fileDiff.newMode) {
    lines.push(`old mode ${fileDiff.oldMode}`);
    lines.push(`new mode ${fileDiff.newMode}`);
  }

  if (fileDiff.isBinary) {
    lines.push(`Binary files a/${fileDiff.oldPath} and b/${fileDiff.newPath} differ`);
    return lines.join('\n');
  }

  lines.push(`--- ${fileDiff.isNew ? '/dev/null' : 'a/' + fileDiff.oldPath}`);
  lines.push(`+++ ${fileDiff.isDeleted ? '/dev/null' : 'b/' + fileDiff.newPath}`);

  for (const hunk of fileDiff.hunks) {
    lines.push(`@@ -${hunk.oldStart},${hunk.oldCount} +${hunk.newStart},${hunk.newCount} @@`);
    lines.push(...hunk.lines);
  }

  return lines.join('\n');
}

export function diffStrings(oldContent: string, newContent: string, filename: string): FileDiff {
  const oldLines = oldContent ? oldContent.split('\n') : [];
  const newLines = newContent ? newContent.split('\n') : [];

  // Check for binary content
  if (isBinary(oldContent) || isBinary(newContent)) {
    return {
      oldPath: filename,
      newPath: filename,
      hunks: [],
      isBinary: true
    };
  }

  const edits = myersDiff(oldLines, newLines);
  const hunks = createHunks(edits);

  return {
    oldPath: filename,
    newPath: filename,
    hunks,
    isNew: oldContent === '',
    isDeleted: newContent === ''
  };
}

function isBinary(content: string): boolean {
  // Check for null bytes or other binary indicators
  if (!content) return false;
  return content.includes('\0');
}

export function diffFiles(oldContent: string, newContent: string, options: {
  oldPath: string;
  newPath: string;
  oldMode?: string;
  newMode?: string;
}): FileDiff {
  // Check for binary content
  if (isBinary(oldContent) || isBinary(newContent)) {
    return {
      oldPath: options.oldPath,
      newPath: options.newPath,
      oldMode: options.oldMode,
      newMode: options.newMode,
      hunks: [],
      isBinary: true
    };
  }

  const oldLines = oldContent ? oldContent.split('\n') : [];
  const newLines = newContent ? newContent.split('\n') : [];

  const edits = myersDiff(oldLines, newLines);
  const hunks = createHunks(edits);

  return {
    oldPath: options.oldPath,
    newPath: options.newPath,
    oldMode: options.oldMode,
    newMode: options.newMode,
    hunks,
    isNew: oldContent === '',
    isDeleted: newContent === ''
  };
}
